Fix mentioning Slack users/groups

aknuds1 opened the pull request:
**What this PR does / why we need it**:
Fix mentioning users and groups in Slack notifications. The current mentioning doesn't work, as you need special Markdown syntax in order to mention users and groups with the current Slack API. You also [need to refer to IDs](https://api.slack.com/changelog/2017-09-the-one-about-usernames) instead of user/group handles.

TODO: Check what happens when Json has an old Mention field

**Which issue(s) this PR fixes**:
Fixes #19938.

aknuds1 committed:
notifiers/slack: Fix mentioning users

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into bugfix/20889

aknuds1 committed:
alerting/slack: Allow mentioning users and groups separately

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

marefr commented on pkg/services/alerting/notifiers/slack.go:
```diff
@@ -73,25 +75,36 @@ func init() {
         </input>
         <info-popover mode="right-absolute">
-          Mention a user or a group using @ when notifying in a channel
+          Mention a user when notifying in a channel
```
Need to extend this I think to include that you must reference user id's and how to include multiple

aknuds1 replied:
@marefr Oh yeah thanks, slipped my mind.

aknuds1 replied:
@marefr I didn't think of supporting multiple user IDs, what should we use for separator? Space, comma?

marefr commented on pkg/services/alerting/notifiers/slack.go:
```diff
@@ -73,25 +75,36 @@ func init() {
+          data-placement="right">
+        </input>
+        <info-popover mode="right-absolute">
+          Mention a group when notifying in a channel
```
Need to extend this I think to include that you must reference user id's and how to include multiple

aknuds1 replied:
@marefr Thanks, slipped my mind.

marefr commented on pkg/services/alerting/notifiers/slack_test.go:
```diff
@@ -51,19 +51,21 @@ func TestSlackNotifier(t *testing.T) {
           "icon_url": "https://grafana.com/img/fav32.png",
-          "mention": "@carl",
+          "mentionUser": "user",
+          "mentionGroup": "group",
```
Would make sense with a test showing multiple user/group mentions

marefr reviewed (changes requested):
Looks good. Some minor things, see comments.

Needs to update docs:
- https://grafana.com/docs/grafana/latest/alerting/notifications/#slack
- https://grafana.com/docs/grafana/latest/administration/provisioning/#alert-notification-slack

papagian commented on pkg/services/alerting/notifiers/slack.go:
```diff
@@ -174,9 +190,20 @@ func (sn *SlackNotifier) Notify(evalContext *alerting.EvalContext) error {
+	}
+	mentionUser := strings.TrimSpace(sn.MentionUser)
+	if mentionUser != "" {
+		message += fmt.Sprintf("<@%s>", mentionUser)
```
Previously we [assumed](https://grafana.com/docs/grafana/latest/alerting/notifications/#slack) that `@` was included in the user input. I think that it will be confusing and prone to errors if we make the users change usernames to IDs and at the same time remove the `@` character from their existing notifiers.

In any case the existing documentation should be updated too and probably we need to add a note to the next release for letting the users fix their notifiers.

aknuds1 replied:
@papagian My thinking at least is that this is such a radical change, splitting user and group mentions (whereas they used to be one). Also, the `@` sign only works for user mentions now, since the character technically happens to be part of the notification syntax. It doesn't work for group mentions, due to its different syntax, so isn't it just more confusing if you have to use `@` for user mentions and not for group mentions?

aknuds1 replied:
I haven't thought about other types of mentions though, than user or group, since the old tooltip just describes those types of mentions. Not sure how many types of mention targets we should support.

papagian replied:
probably yes! so just make sure we communicate this properly to the users.

papagian reviewed (changes requested):
We need to update the documentation.

aknuds1 committed:
Support channel mentions

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Support mentioning multiple users/groups

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Improve tests

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Update documentation

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

marefr commented on docs/sources/alerting/notifications.md:
```diff
@@ -75,20 +75,22 @@ able to access the image.
-Mention | make it possible to include a mention in the Slack notification sent by Grafana. Ex @here or @channel
+Mention Users | Mention one or more comma separated users in the Slack notification sent by Grafana. You have to refer to users via their Slack IDs (you can get a user's ID from their Slack profile).
+Mention Groups | Mention one or more comma separated groups in the Slack notification sent by Grafana. You have to refer to groups via their Slack IDs (you can get a group's ID from its Slack profile URL).
+Mention Channel | Mention either all channel members (_@channel_) or just active ones (_@here_).
```
Maybe better UX to have a drop down with options `Disabled`, `Everyone`, `Every online member`.

marefr commented on docs/sources/alerting/notifications.md:
```diff
@@ -75,20 +75,22 @@ able to access the image.
 Icon emoji | Provide an emoji to use as the icon for the bot's message. Ex :smile:
 Icon URL | Provide a url to an image to use as the icon for the bot's message.
-Mention | make it possible to include a mention in the Slack notification sent by Grafana. Ex @here or @channel
+Mention Users | Mention one or more comma separated users in the Slack notification sent by Grafana. You have to refer to users via their Slack IDs (you can get a user's ID from their Slack profile).
```
Could you maybe include an example url, I'm looking in the slack app but cannot find any id. Do you have to go to website/administration of slack workspace? Same in a few different places that makes me confused.

aknuds1 replied:
@marefr I'm not sure what you mean by example URL. It's sort of difficult to explain more I think, without going really into detail with screenshots and all. Basically you have to f.ex. right click a username in chat and click their name to visit their profile, from where you can click the "More" button and then "Copy member ID".

I don't think it makes sense to be too detailed here about how to navigate Slack's UI, since it could change and then our documentation is stale.

marefr commented on pkg/services/alerting/notifiers/slack.go:
```diff
@@ -110,37 +137,61 @@ func NewSlackNotifier(model *models.AlertNotification) (alerting.Notifier, error
-	Username  string
-	IconEmoji string
-	IconURL   string
-	Mention   string
```
Think we need to document this as a breaking change as well.

aknuds1 replied:
@marefr Is there anything I need to do in this regard?

bergquist replied:
The breaking change is in Slacks API but our users is still affected so I think it might be a good idea to add a note about this in the changelog.

aknuds1 replied:
@bergquist OK, adding an item to the changelog.

aknuds1 replied:
@marefr @bergquist Added an item to 6.7.0 breaking changes in changelog. Please let me know if I should do it differently.

marefr reviewed (changes requested):
Some minor things left. See comments. 

Found some proper documentation of the mentioning formatting: https://api.slack.com/reference/surfaces/formatting#advanced

aknuds1 committed:
Change channel mention options according to feedback

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Tweak docs

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into bugfix/20889

aknuds1 committed:
CHANGELOG: Add Slack breaking change

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

marefr committed:
Docs: Document configuration of console, file and syslog log formats (#21768)

Adds missing configuration documentation of the different 
log formats; console, file and syslog.

Ref #21765

hugohaggmark committed:
Chore: Removes Cypress record (#21782)

hugohaggmark committed:
Chore: Lowers strict error limit (#21781)

bergquist committed:
Add link guide for installing new renderer (#21702)

papagian commented on pkg/services/alerting/notifiers/slack.go:
```diff
@@ -73,25 +75,50 @@ func init() {
+          data-placement="right">
+		  <option value="">Disabled</option>
+		  <option value="here">Every active channel member</option>
+		  <option value="channel">Every channel member</option>
```
there is an additional `everyone` [option](https://api.slack.com/reference/surfaces/formatting#special-mentions) that notifies every person in the #general channel (ie. every non-guest member of a workspace).

aknuds1 replied:
@papagian Yeah, but I didn't really think supporting `@everyone` really made sense since this notification is for a channel. My reasoning was that if you are notifying a channel, why would you also want to notify `#general`? I could be convinced I'm wrong though, maybe there are valid use cases?

papagian replied:
ok, you have convinced me :)

papagian commented:
I suspect that the `Recipient` field also [expects](https://github.com/grafana/grafana/blob/f613214ac5633bf5aefb585bea3e39276b25e9d6/pkg/services/alerting/notifiers/slack.go#L41) group IDs or user IDs.

aknuds1 commented:
> I suspect that the `Recipient` field also [expects](https://github.com/grafana/grafana/blob/f613214ac5633bf5aefb585bea3e39276b25e9d6/pkg/services/alerting/notifiers/slack.go#L41) group IDs or user IDs.

Topic for another PR?

papagian commented:
> > I suspect that the `Recipient` field also [expects](https://github.com/grafana/grafana/blob/f613214ac5633bf5aefb585bea3e39276b25e9d6/pkg/services/alerting/notifiers/slack.go#L41) group IDs or user IDs.
> 
> Topic for another PR?

It could be, but in that case we need to notify the users to update their rules once again.

aknuds1 commented:
> > > I suspect that the `Recipient` field also [expects](https://github.com/grafana/grafana/blob/f613214ac5633bf5aefb585bea3e39276b25e9d6/pkg/services/alerting/notifiers/slack.go#L41) group IDs or user IDs.
> > 
> > 
> > Topic for another PR?
> 
> It could be, but in that case we need to notify the users to update their rules once again.

Only if the change happens in a different release?

ivanahuckova committed:
Logs panel: Rename labels to unique labels (#21783)

papagian commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 # 6.7.0 (unreleased)
 
 ## Breaking changes
+* **Slack**: Remove Mention setting and instead introduce _Mention Users_, _Mention Groups_, and _Mention Channel_. The first two settings require user and group IDs respectively. This change was necessary due to the fact that the way of mentioning via the Slack API changed, and mentions in Slack notifications no longer worked.
```
You could add a link mentioning the modification in the Slack API.

aknuds1 replied:
@papagian Done.

aknuds1 committed:
CHANGELOG: Add link to breaking change

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

sakjur committed:
Config: add meta feature toggle (#21786)

sv5d committed:
Heatmap: Legend color range is incorrect when using custom min/max (#21748)

Fix opacity and color scales initializations which took wrong range 
arguments (current data min/max instead of custom range min/max).

Fixes #18139

marefr committed:
OpenTSDB: Add back missing ngInject (#21796)

Adds back missing ngInject on datasource constructor
to make it work again.

Fixes #21770

xlson committed:
MSI: License for Enterprise (#21794)

marefr committed:
Devenv: OpenTSDB dashboard (#21797)

Simple test dashboard for OpenTSDB.

aocenas committed:
Influxdb: Fix cascader when doing log query in explore (#21787)

* Fix cascader options and add tests

* Add comment

* Fix typo

oddlittlebird commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 # 6.7.0 (unreleased)
 
 ## Breaking changes
+* **Slack**: Remove Mention setting and instead introduce _Mention Users_, _Mention Groups_, and _Mention Channel_. The first two settings require user and group IDs respectively. This change was necessary due to the fact that the way of mentioning via the Slack API [changed](https://api.slack.com/changelog/2017-09-the-one-about-usernames) and mentions in Slack notifications no longer worked.
```
```suggestion
* **Slack**: Removed _Mention_ setting and instead introduce _Mention Users_, _Mention Groups_, and _Mention Channel_. The first two settings require user and group IDs, respectively. This change was necessary because the way of mentioning via the Slack API [changed](https://api.slack.com/changelog/2017-09-the-one-about-usernames) and mentions in Slack notifications no longer worked.
```

oddlittlebird commented on docs/sources/alerting/notifications.md:
```diff
@@ -75,20 +75,22 @@ able to access the image.
 {{< imgbox max-width="40%" img="/img/docs/v4/slack_notification.png" caption="Alerting Slack Notification" >}}
 
-To set up slack you need to configure an incoming webhook url at slack. You can follow their guide on how
+To set up Slack, you need to configure an incoming Slack webhook URL. You can follow their guide on how
```
```suggestion
To set up Slack, you need to configure an incoming Slack webhook URL. You can follow the Slack guide on how
```

aknuds1 replied:
To be honest, I think the older wording is better, since the new wording sounds overly repetitive to my ears (3x Slack in the same sentence). Besides, wording wasn't introduced by my PR, so if you want to change it around I think it's better to make a new PR.

oddlittlebird commented on docs/sources/alerting/notifications.md:
```diff
@@ -75,20 +75,22 @@ able to access the image.
 
-To set up slack you need to configure an incoming webhook url at slack. You can follow their guide on how
+To set up Slack, you need to configure an incoming Slack webhook URL. You can follow their guide on how
 to do that [here](https://api.slack.com/incoming-webhooks). If you want to include screenshots of the firing alerts
```
Please rewrite this so that it removes "here" as the link text. Per the Documentation Style Guide, we want to use descriptive linking text, such as the title of the page being linked to, rather than uninformative verbiage such as "here."

aknuds1 replied:
@oddlittlebird The text in question isn't touched by my PR so I think it's out of scope for the review, but since it's such an uncontroversial change, sure.

oddlittlebird commented on docs/sources/alerting/notifications.md:
```diff
@@ -75,20 +75,22 @@ able to access the image.
 Icon emoji | Provide an emoji to use as the icon for the bot's message. Ex :smile:
 Icon URL | Provide a url to an image to use as the icon for the bot's message.
-Mention | make it possible to include a mention in the Slack notification sent by Grafana. Ex @here or @channel
+Mention Users | Optionally mention one or more comma separated users in the Slack notification sent by Grafana. You have to refer to users via their Slack IDs (you can get a user's ID from their Slack profile).
```
```suggestion
Mention Users | Optionally mention one or more comma-separated Slack IDs in the Slack notification sent by Grafana. You can get a user's ID from their Slack profile.
```

aknuds1 replied:
@oddlittlebird Based on my knowledge of the Slack API specs and my practical usage of it, I think the suggested wording is less technically precise. If I were to read this documentation in order to understand the semantics of the setting, I would definitely find the original wording the most helpful. The reason being that the resulting message mentions users and not IDs; the IDs (and not any kind of Slack IDs, but _user_ IDs in particular) just happen to be how users are referenced when communicating with the Slack API.

oddlittlebird replied:
I was mostly trying to combine the two sentences, since they seemed kind of choppy. What about Slack user IDs? If that does not work for you, then just make sure the hyphen in "comma-separated" is added and I shall be happy.

aknuds1 replied:
@oddlittlebird I changed to "comma-separated already". I think I see how I can improve the sentence. The basic structure of the sentence is necessary, since while the goal is to mention Slack users, for this setting you have to provide user IDs, so you can't just refer to the latter.

aknuds1 replied:
@oddlittlebird I rewrote the sentences a bit, to make it clear that the IDs of the setting must be comma-separated, and that it doesn't apply to the resulting mentions in Slack.

oddlittlebird replied:
Much better, thank you!

oddlittlebird commented on docs/sources/alerting/notifications.md:
```diff
@@ -75,20 +75,22 @@ able to access the image.
 Icon URL | Provide a url to an image to use as the icon for the bot's message.
-Mention | make it possible to include a mention in the Slack notification sent by Grafana. Ex @here or @channel
+Mention Users | Optionally mention one or more comma separated users in the Slack notification sent by Grafana. You have to refer to users via their Slack IDs (you can get a user's ID from their Slack profile).
+Mention Groups | Optionally mention one or more comma separated groups in the Slack notification sent by Grafana. You have to refer to groups via their Slack IDs (you can get a group's ID from its Slack profile URL).
```
```suggestion
Mention Groups | Optionally mention one or more comma-separated group Slack IDs in the Slack notification sent by Grafana. You can get a group's ID from its Slack profile URL.
```

aknuds1 replied:
See my explanation above.

oddlittlebird commented on docs/sources/alerting/notifications.md:
```diff
@@ -75,20 +75,22 @@ able to access the image.
-Mention | make it possible to include a mention in the Slack notification sent by Grafana. Ex @here or @channel
+Mention Users | Optionally mention one or more comma separated users in the Slack notification sent by Grafana. You have to refer to users via their Slack IDs (you can get a user's ID from their Slack profile).
+Mention Groups | Optionally mention one or more comma separated groups in the Slack notification sent by Grafana. You have to refer to groups via their Slack IDs (you can get a group's ID from its Slack profile URL).
+Mention Channel | Optionally mention either all channel members or just active ones.
```
How do I choose if it will mention all channel members or just active ones?

aknuds1 replied:
@oddlittlebird It's self explanatory through the UI options (you can choose "Every active channel member" or "Every channel member").

Commit:
Azure Monitor: Fix Application Insights API key field to allow input (#21738)

* Fix update api key input

* update snapshot

hugohaggmark committed:
Fix: Fixes user logout for datasourceRequests with 401 from  datasource (#21800)

tskarhed committed:
Select: Fix scroll issue (#21795)

mckn committed:
Fix: prevents the BarGauge from exploding when the datasource returns empty result. (#21791)

* Fixed issue where gauge throw error on empty result.

* Some refactorings to improve the code.

* Added some tests to make sure this doesn't happen again.

aknuds1 committed:
Update CHANGELOG.md

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

aknuds1 committed:
docs: Use descriptive link

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Docs: Tweak language

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into bugfix/20889

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into bugfix/20889

aknuds1 committed:
Tweak docs

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into bugfix/20889

kylebrandt commented on pkg/services/alerting/notifiers/slack.go:
```diff
@@ -174,9 +225,32 @@ func (sn *SlackNotifier) Notify(evalContext *alerting.EvalContext) error {
 	}
 
-	message := sn.Mention
+	message := ""
```
Should probably be a string builder? Also maybe maybe something like:

```
	foo := ""
	msgAddSpace := func() {
		if foo != "" {
			foo += " "
		}
		return
	}
	msgAddSpace()
``` 

to reduce repetition,  bit with stringbuilder.

aknuds1 replied:
Thanks @kylebrandt - will look into it.

aknuds1 committed:
Use strings.Builder

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

marefr commented on pkg/services/alerting/notifiers/slack.go:
```diff
@@ -185,13 +258,21 @@ func (sn *SlackNotifier) Notify(evalContext *alerting.EvalContext) error {
 	}
 
 	body := map[string]interface{}{
+		"blocks": []map[string]interface{}{
```
What's the reason for moving message into a block which seem to use `mrkdwn` format? Seems like notifications cannot show preview of message any longer:
![image](https://user-images.githubusercontent.com/1668778/73825307-089c1980-47fc-11ea-970b-56a60e895274.png)
![image](https://user-images.githubusercontent.com/1668778/73825593-88c27f00-47fc-11ea-92d9-de541bd6d332.png)

marefr replied:
Need to verify that it showed preview before

marefr replied:
On master I see 
![image](https://user-images.githubusercontent.com/1668778/73826016-536a6100-47fd-11ea-88a8-23b3aff1071d.png)

aknuds1 replied:
@marefr It's because, based on my understanding of Slack's API specs, you need to use an attachment with Markdown in order to mention users/groups/channels. I'll read more closely though, to see if you should e.g. have a fallback message.

aknuds1 replied:
@marefr I fixed the issue. The API specs are really weak on explaining this (I couldn't find the behaviour documented), but it looks as if you need to include a top-level `text` field in your message object, with just plain text (no markup) as a fallback for Web and mobile notifications. You also need the block object in order to deliver mention markup, since that doesn't work in the `text` field.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into bugfix/20889

aknuds1 committed:
Include raw text message for fallback

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

papagian commented on pkg/services/alerting/notifiers/slack_test.go:
```diff
@@ -51,19 +51,23 @@ func TestSlackNotifier(t *testing.T) {
-          "mention": "@carl",
+          "mentionUsers": "user1, user2",
+          "mentionGroups": "group1, group2",
+		  "mentionChannel": "here",
```
Can you fix indentation here?

aknuds1 committed:
Fix indentation

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

marefr commented on pkg/services/alerting/notifiers/slack.go:
```diff
@@ -185,13 +258,22 @@ func (sn *SlackNotifier) Notify(evalContext *alerting.EvalContext) error {
 	}
 
 	body := map[string]interface{}{
+		"text": evalContext.Rule.Message,
```
I think you want to use `evalContext.GetNotificationTitle()` here instead to make it backward compatible.

marefr reviewed (changes requested):
Use rule title instead of message, see comment.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into bugfix/20889

aknuds1 committed:
Use notification title as fallback

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Remove fallback from Slack attachment

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into bugfix/20889

aknuds1 committed:
Put message back in attachment

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented:
@marefr I put the message back in attachments, only appending a block if there are mentions. Please have another look.

marefr reviewed (approved):
Works as expected. Great!

LGTM

papagian reviewed (approved):
LGTM
